pool.go

```diff
@@ -239,6 +245,12 @@ func (p *Pool) Start(closeFunc func()) {
 			}
 			log.Info("finish closing pool")
 
+			// wait for all on-fly retryable jobs are inserted to db
+			// prevent negative waitgroup by checking numberOfRetryingJob
+			if atomic.LoadInt32(&p.numberOfRetryingJob) > 0 {
+				p.retryableWaitGroup.Wait()
```

Comment:
I think the negative waitgroup happens when we call Done more than we call Add. This is the case of previous commit as `updateRetryingJob` can be called directly in Pool.Start in case p.ctx.Done() which does not have Add called. Due to the specification, `Add(1)` cannot be called concurrently with `Wait` when the counter is 0, so the atomic counter here may prevent that problem, but I feel like using another atomic counter is not the correct way as waitgroup already uses atomic counter, maybe we can look at semaphore.
Theoretically, with the current order
```golang
atomic.AddInt32(&p.numberOfRetryingJob, 1)
p.retryableWaitGroup.Add(1)
```
The pool.Start falls into RetryJobChan -> PrepareRetryableJob -> ctx.Done -> numberOfRetryingJob == 0 (don't wait) -> PrepareRetryableJob: Add(1) + store DB
So the pool may not wait for the completion of DB store, but we cannot swap the order as this can cause `Add(1)`  be called concurrently with `Wait` when the counter is 0. This is such a small window so maybe we can think about this later.